### DISCUSSION TITLE
Decoding issue

### DISCUSSION THREAD

**Discussion by saroar** [pointfreeco/swift-url-routing#48] [Q&A]
My playground code which dont have any issue 
playground code here https://gist.github.com/saroar/a1bab728f4292c54cb76085a633c6a3c
my client iOS app dont send any request to server becz its not decoding my query 
https://github.com/AddaMeSPB/AddaMeIOS/blob/siteRouter/AddameSPM/Sources/EventView/EventsReducer.swift#L112
error coming from decoding from this 
https://github.com/pointfreeco/swift-url-routing/blob/main/Sources/URLRouting/Client/Client.swift#L53

```swift
HTTPRequest.HRError(
        description: "fetch events get error",
        reason: URLRoutingDecodingError(
          bytes: Data(35 bytes),
          response: NSHTTPURLResponse(),
          underlyingError: DecodingError.dataCorrupted(
            DecodingError.Context(
              codingPath: [],
              debugDescription: "The given data was not valid JSON.",
              underlyingError: NSError(
                domain: "NSCocoaErrorDomain",
                code: 3840,
                userInfo: [
                  "NSDebugDescription": "Invalid value around line 1, column 0.",
                  "NSJSONSerializationErrorIndex": 0
                ]
              )
            )
          )
        )
      )
```

**Comment by saroar** [comment]
<img width="1163" alt="Screenshot 2022-08-30 at 08 45 59" src="https://user-images.githubusercontent.com/8770772/187359801-13300b30-6d82-42b0-99b1-7415c580ed75.png">

**Comment by saroar** [comment]
I have test it on my api server working good only issue was from iOS client 
https://github.com/AddaMeSPB/EventEngine/blob/master/Tests/AppTests/Event/EventsRoutingTests.swift#L14

**Comment by saroar** [comment]
I am getting this decoding issue lots of time I am not sure its have issue or dont get proper error message becz 

```swift
  case let .imageUploadResponse(.success(imageURLString)):
        guard let user: UserOutput = KeychainService.loadCodable(for: .user) else {
          return .none
        }

      let attachment = AttachmentInOutPut(type: .image, userId: user.id, imageUrlString: imageURLString)
      
      return .task {
          do {
              
              let attacment = try await environment.attachmentClient.updateUserImageURL(attachment)
              return ProfileAction.attacmentResponse(.success(attacment))
          } catch {
              return ProfileAction.attacmentResponse(
                .failure(
                    HTTPRequest.HRError.custom("cant upload attachment", error)
                )
              )
          }
      }
``` 

from clicnt I am trying to send like this I am 100% sure i dont have decoding issue at  all 

here my client https://github.com/AddaMeSPB/AddaMeIOS/blob/siteRouter/AddameSPM/Sources/AttachmentClient/AttachmentClient.swift
here my live implement https://github.com/AddaMeSPB/AddaMeIOS/blob/siteRouter/AddameSPM/Sources/AttachmentClientLive/Live.swift#L91

my reducer https://github.com/AddaMeSPB/AddaMeIOS/blob/siteRouter/AddameSPM/Sources/ProfileView/ProfileReducer.swift#L50

and form this https://github.com/AddaMeSPB/AddaMeIOS/blob/siteRouter/AddameSPM/Sources/AttachmentClientLive/Live.swift#L84
I am getting sometimes error too 

```swift
HRError
  - description : "Failed to load the request: Optional(HTTPClientError.deadlineExceeded)"
  ▿ reason : Optional<Error>
    ▿ some : HTTPClientError.deadlineExceeded
      - code : AsyncHTTPClient.HTTPClientError.(unknown context at $104271798).Code.deadlineExceeded
```


testing from server side no issue 
https://github.com/AddaMeSPB/AddaMeAuth/blob/siteRouter/Tests/AppTests/AttachmentTests.swift#L26

**Comment by stephencelis** [comment]
@saroar It looks like the server is sending back a "deadline exceeded" error, which is why decoding is failing.

You can see the definition here:

https://github.com/swift-server/async-http-client/blob/7f998f5118de081fa2034833ce95bb5925dc6459/Sources/AsyncHTTPClient/HTTPClient.swift#L1181-L1182

I'm not familiar with this error or how to work around it, but I don't think it's a bug with our library. Can you ask someone in a community more familiar with the `HTTPClient` library, and whatever server code you are using to serve up that response? They may be able to help you debug. If it turns out to be an issue with our library we'd be happy to address! We just need a reproduction.

Since this doesn't appear to be a bug in our library, I'm going to close it. But if you do believe there's a bug in our library, please file a new issue with a reproduction for us to work with and instructions on how to reproduce. Without steps to reproduce, we really don't have much of an ability to try to fix.

**Reply by saroar** [reply]
thanks for reply @stephencelis but we had two error as also I send you test from server without any issue have you check that one?
for this https://github.com/AddaMeSPB/AddaMeAuth/blob/siteRouter/Tests/AppTests/AttachmentTests.swift#L26 from client I have decode issue too but test it passed in server side so please check it I have post all code which you can check

**Reply by saroar** [reply]
also note that I dont have this kind of issue at all before I am greeting this issue after using two SPM
1. VaporRouting
2. URLRouting
what I am thinking it happening becz request was cancelling but why I dont know for now 
I THINK ISSUE IS WHEN YOU HAVE LOW INTERNET SO IT HAPPEN TIME OUT OR DISCONNECT FROM CLEINT

**Reply by stephencelis** [reply]
@saroar It sounds like this is an error that you should be handling in your application. `URLRoutingDecodingError` is public, so you could do the following:

```swift
do {
  let response = try await client.decodedResponse(for: .route, as: MyResponse.self)
  // ...
} catch let error as URLRoutingDecodingError {
  // use error.response or error.data to surface errors to user
} catch {
  // any other error handling
}
```

**Reply by saroar** [reply]
I see let me try it before I like to share this I use for send big string before 
```swift
extension RoutesBuilder {
    @discardableResult
    public func postBigFile<Response>(
        _ path: PathComponent...,
        use closure: @escaping (Request) throws -> Response
    ) -> Route
        where Response: ResponseEncodable
    {
        return self.on(.POST, path, body: .collect(maxSize: 50_000_000), use: closure)
    }

    @discardableResult
    public func postBigFile<Response>(
        _ path: [PathComponent],
        use closure: @escaping (Request) throws -> Response
    ) -> Route
        where Response: ResponseEncodable
    {
        return self.on(.POST, path, body: .collect(maxSize: 50_000_000), use: closure)
    }

    @discardableResult
    public func putBigFile<Response>(
        _ path: PathComponent...,
        use closure: @escaping (Request) throws -> Response
    ) -> Route
        where Response: ResponseEncodable
    {
        return self.on(.PUT, path, body: .collect(maxSize: 10_000_000), use: closure)
    }

    @discardableResult
    public func putBigFile<Response>(
        _ path: [PathComponent],
        use closure: @escaping (Request) throws -> Response
    ) -> Route
        where Response: ResponseEncodable
    {
        return self.on(.PUT, path, body: .collect(maxSize: 10_000_000), use: closure)
    }
}
```

this is I am sending 
https://gist.github.com/saroar/a12b52c532d1d4a09879a9fe9d05a42e


here after I try your code 
debugPrint(#line, error.response, error.localizedDescription)
```
79 <NSHTTPURLResponse: 0x28005c020> { URL: http://10.0.1.4:8080/v1/contacts } { Status Code: 404, Headers {
    Connection =     (
        "keep-alive"
    );
    "Content-Length" =     (
        359
    );
    Date =     (
        "Fri, 09 Sep 2022 14:51:06 GMT"
    );
} } 

"The operation couldn’t be completed. (URLRouting.URLRoutingDecodingError error 1.)"
```

// from server its very strange 
[ INFO ] POST /v1/contacts [request-id: D83347F2-65B8-466D-850B-B415CB784908]
[ INFO ] error: multiple failures occurred

error: unexpected input
 --> input:1:5
1 | /v1/contacts
  |     ^ expected "events"
  |     ^ expected "categories"
  |     ^ expected "conversations"
  |     ^ expected "messages"
  |     ^ expected "users"

error: RoutingError()

error: unexpected input
 --> input:1:5
1 | /v1/contacts
  |     ^ expected "devices" [request-id: D83347F2-65B8-466D-850B-B415CB784908]

NOTE: when I send this request from my wife phone all work good (iPhone 13 pro ) also work from simulator 
may be base I am seining less data

**Reply by saroar** [reply]
this is work with simulator just wanted to share with you 

https://user-images.githubusercontent.com/8770772/189381760-fcd60deb-6bb4-449a-aff3-adc9107b587d.mp4

**Reply by saroar** [reply]
@stephencelis I am still stuck with this can you help me find out what is going on? I write server test all good in server only issue from IOS client :(

**Reply by stephencelis** [reply]
@saroar Unfortunately it's really difficult for me to troubleshoot your entire stack 😕 The error you show above has it coming from the server, though, which makes me think it's not a bug in our library, and I'm just not familiar enough with the rest of your stack to debug further, sorry!

Just some questions I'd ask myself to see if it helps: Is the error consistent? Or does it only happen occasionally (you mentioned your internet connection earlier)? When you get an error, try to manually inspect the request that led to it and see if anything looks off, or if making any changes to that request leads to something that works better. Good luck!

**Reply by saroar** [reply]
but I send you like where I am testing on back end without any issue I am pass test only problem is from some real device work some not 
from which device it was not working I grab same input then try to run test on back end all test passed 
I also send u video where some real device working 

testing from server side no issue
https://github.com/AddaMeSPB/AddaMeAuth/blob/siteRouter/Tests/AppTests/ContactTests.swift#L21

**Reply by saroar** [reply]
@stephencelis
